### PR TITLE
[docs] LibraryEvolution: Tweak for Swift 4.0

### PR DESCRIPTION
**Pull request by jrose-apple**
Most interesting changes:
- Update for generic subscripts.
- Disallow reordering stored instance properties for a fixed-contents struct. We figured people want to rename non-public properties more than they want to reorder them, and we can only use the name or the ordering to provide a stable tie-breaker for layout.

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
I'm actually not quite sure my ordering thing is correct, since in theory you could use the type of a property as a partial tie-breaker as well.

**Comment by jrose-apple** [comment]
Okay, I've convinced myself that it *isn't* correct to allow both reordering and renaming, because the current rules allow taking a private field and making it public. Even though no one can depend on its position until it becomes public, it has to not change the offset of *other* fields.

**Comment by jrose-apple** [comment]
(We originally thought we could sort by availability here, but that doesn't work during *development* of a new version of a library, where first one property is made public and then another. You'd potentially have to number every single build for that to work.)

File: docs/LibraryEvolution.rst [hunk 1]
```diff
@@ -932,8 +912,8 @@ little more restrictive than structs; they only allow the following changes:
 
 Finally, classes allow the following changes that do not apply to structs:
 
-- A class may be marked ``open`` if it is not already marked ``final``.
-- A class may be marked ``final`` if it is not already marked ``open``.
+- A public class may be made ``open`` if it is not already marked ``final``.
+- A non-``open`` public class may be marked ``final``.
 - Removing an explicit deinitializer. (A class with no declared deinitializer
```
Comment (slavapestov) [inline docs/LibraryEvolution.rst]: Any reason adding an explicit deinitializer is not safe?

**Comment by slavapestov** [comment]
I think struct layout needs more thought but the other changes LGTM.

**Review by slavapestov — APPROVED** [review]

File: docs/LibraryEvolution.rst (hunk 1 above)
Comment (jrose-apple) [reply docs/LibraryEvolution.rst]: Whoops, yes, that should be there too. I wonder how I missed that the first time.

(Or maybe I was worried about us using it to optimize, but you already can't do that if you can add members.)

**Comment by jrose-apple** [comment]
Yeah, maybe I'll bring it up on swift-dev.

**Pull request merged**